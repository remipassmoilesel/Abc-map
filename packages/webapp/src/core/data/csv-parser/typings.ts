/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

export declare type CsvRow = { [k: string]: string | number | undefined };

export const UnknownLineNumber = -1;

export class CsvParsingError extends Error {
  constructor(message: string, public line: number) {
    super(message);
  }
}

export function isCsvParsingError(err: unknown): err is CsvParsingError {
  return !!err && typeof err === 'object' && 'line' in err && typeof err.line !== 'undefined';
}
