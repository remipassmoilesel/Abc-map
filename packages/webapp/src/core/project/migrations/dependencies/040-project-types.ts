/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { AbcLayer, AbcLayout } from '@abc-map/shared';

export interface AbcProjectManifest040 {
  metadata: AbcProjectMetadata040;
  layers: AbcLayer[];
  layouts: AbcLayout[];
}

export interface AbcProjectMetadata040 {
  id: string;
  version: string;
  name: string;
  containsCredentials: boolean;
}

export interface LayoutFormat040 {
  name?: string;
  id: string;
  width: number;
  height: number;
  orientation: 'portrait' | 'landscape';
}
