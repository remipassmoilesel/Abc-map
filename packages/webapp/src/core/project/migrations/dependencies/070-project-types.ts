/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { AbcLayout, AbcSharedView } from '@abc-map/shared';
import { AbcLegend060 } from './060-legend-types';

export interface AbcProjectManifest070 {
  metadata: AbcProjectMetadata070;
  layouts: AbcLayout070[];
  sharedViews: AbcSharedView070[];
  view: any;
  layers: any[];
}

export interface AbcLayout070 extends AbcLayout {
  legend: AbcLegend060;
}

export interface AbcSharedView070 extends AbcSharedView {
  legend: AbcLegend060;
}

export interface AbcProjectMetadata070 {
  id: string;
  version: string;
  name: string;
  public: boolean;
  containsCredentials: boolean;
}
