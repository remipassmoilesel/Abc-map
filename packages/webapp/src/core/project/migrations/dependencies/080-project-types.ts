/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { AbcSharedView } from '@abc-map/shared';

export interface AbcProjectManifest080 {
  metadata: AbcProjectMetadata080;
  sharedViews: AbcSharedView[];
  layouts: any[];
  view: any;
  layers: any[];
}

export interface AbcProjectMetadata080 {
  id: string;
  version: string;
  name: string;
  public: boolean;
  containsCredentials: boolean;
}
