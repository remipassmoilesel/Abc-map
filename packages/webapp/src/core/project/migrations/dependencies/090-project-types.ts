/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { AbcLayer, AbcLayout, AbcSharedView, AbcView } from '@abc-map/shared';

export interface AbcProjectManifest090 {
  metadata: AbcProjectMetadata090;
  sharedViews: {
    list: AbcSharedView[];
    fullscreen: boolean;
    mapDimensions: { width: number; height: number };
  };
  layouts: AbcLayout[];
  view: AbcView;
  layers: AbcLayer[];
}

export interface AbcProjectMetadata090 {
  id: string;
  version: string;
  name: string;
  public: boolean;
  containsCredentials: boolean;
}
