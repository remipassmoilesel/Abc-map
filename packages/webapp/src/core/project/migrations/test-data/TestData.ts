/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import * as fs from 'fs';
import { MigrationProject } from '../typings';
import { AbcProjectManifest, ProjectHelper, Zipper } from '@abc-map/shared';

export class TestData {
  public static async project01(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.1.abm2`);
  }

  public static async project020(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.2.0.abm2`);
  }

  public static async project030(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.3.0.abm2`);
  }

  public static async project040(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.4.0.abm2`);
  }

  public static async project050(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.5.0.abm2`);
  }

  public static async project060(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.6.0.abm2`);
  }

  public static async project070(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.7.0.abm2`);
  }

  public static async project080(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.8.0.abm2`);
  }

  public static async project090(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-0.9.0.abm2`);
  }

  public static async project100(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-1.0.0.abm2`);
  }

  public static async project110(): Promise<MigrationProject> {
    return this.readFile(`${__dirname}/project-1.1.0.abm2`);
  }

  public static fakeProject(version: string) {
    return { metadata: { version } } as unknown as AbcProjectManifest;
  }

  private static async readFile(path: string): Promise<MigrationProject> {
    const archive = fs.readFileSync(path);
    const files = await Zipper.forNodeJS().unzip(archive);
    const manifest = await ProjectHelper.forNodeJS().extractManifest(archive);
    const blobFiles = files.map((f) => ({ path: f.path, content: new Blob([f.content as Buffer]) }));

    return { manifest, files: blobFiles };
  }
}
