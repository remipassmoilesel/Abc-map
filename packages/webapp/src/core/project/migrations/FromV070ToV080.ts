/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcFile, AbcLayout, AbcSharedView, AbcTextFrame, Logger, TableElement, TableRowElement, TextFrameChild } from '@abc-map/shared';
import { MigrationProject, ProjectMigration } from './typings';
import semver from 'semver';
import { AbcLayout070, AbcProjectManifest070, AbcSharedView070 } from './dependencies/070-project-types';
import { AbcLegend060, LegendDisplay } from './dependencies/060-legend-types';
import { nanoid } from 'nanoid';
import { AbcProjectManifest080 } from './dependencies/080-project-types';

const NEXT = '0.8.0';

const logger = Logger.get('FromV070ToV080.ts');

/**
 * This migration transforms legends in text frames
 */
export class FromV070ToV080 implements ProjectMigration<AbcProjectManifest070, AbcProjectManifest080> {
  public async interestedBy(manifest: AbcProjectManifest070): Promise<boolean> {
    const version = manifest.metadata.version;
    return semver.lt(version, NEXT);
  }

  public async migrate(_manifest: AbcProjectManifest070, files: AbcFile<Blob>[]): Promise<MigrationProject<AbcProjectManifest080>> {
    const manifest = _manifest as unknown as AbcProjectManifest070;

    const layouts: AbcLayout[] = manifest.layouts.map((lay) => {
      const frame = this.legendToFrame(lay.legend);
      const migrated = { ...lay, textFrames: frame ? [frame] : [] };
      delete (migrated as Partial<AbcLayout070>).legend;
      return migrated;
    });

    const sharedViews: AbcSharedView[] = manifest.sharedViews.map((view) => {
      const frame = this.legendToFrame(view.legend);
      const migrated = { ...view, textFrames: frame ? [frame] : [] };
      delete (migrated as Partial<AbcSharedView070>).legend;
      return migrated;
    });

    return {
      manifest: {
        ...manifest,
        layouts,
        sharedViews,
        metadata: {
          ...manifest.metadata,
          version: NEXT,
        },
      },
      files,
    };
  }

  private legendToFrame(legend: AbcLegend060): AbcTextFrame | undefined {
    if (legend.display === LegendDisplay.Hidden) {
      return;
    }

    // Create table from items
    const table: TableElement = {
      type: 'table',
      border: 0,
      children: [],
    };

    for (const item of legend.items) {
      const row: TableRowElement = { type: 'table-row', children: [] };

      // Add symbol
      let firstCell: TextFrameChild;
      if (item.symbol) {
        firstCell = { type: 'map-symbol', style: item.symbol.properties, geometryType: item.symbol.geomType, children: [{ text: '' }] };
      } else {
        firstCell = { text: '' };
      }

      row.children.push({ type: 'table-cell', children: [{ type: 'paragraph', align: 'center', children: [firstCell] }] });

      // Add text
      row.children.push({ type: 'table-cell', children: [{ type: 'paragraph', children: [{ text: item.text }] }] });

      table.children.push(row);
    }

    return {
      id: nanoid(),
      size: { width: legend.width * 2, height: legend.height * 2.5 },
      position: { x: 0, y: 0 },
      content: [{ type: 'paragraph', children: [{ text: '' }] }, table, { type: 'paragraph', children: [{ text: '' }] }],
    } as AbcTextFrame;
  }
}
