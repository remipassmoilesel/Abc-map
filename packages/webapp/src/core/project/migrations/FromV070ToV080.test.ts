/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { MigrationProject } from './typings';
import { TestData } from './test-data/TestData';
import { FromV070ToV080 } from './FromV070ToV080';
import { mockNanoid, restoreNanoid } from '../../utils/test/mock-nanoid';
import { AbcProjectManifest070 } from './dependencies/070-project-types';

describe('FromV070ToV080', () => {
  let sampleProject: MigrationProject<AbcProjectManifest070>;
  let migration: FromV070ToV080;

  beforeEach(async () => {
    sampleProject = await TestData.project070();
    migration = new FromV070ToV080();

    mockNanoid();
  });

  afterEach(() => {
    restoreNanoid();
  });

  it('interestedBy() should return true if version < 0.8', async () => {
    expect(await migration.interestedBy(sampleProject.manifest)).toEqual(true);
    expect(await migration.interestedBy(TestData.fakeProject('0.8.0'))).toEqual(false);
  });

  it('migrate should work', async () => {
    // Act
    const result = await migration.migrate(sampleProject.manifest, sampleProject.files);
    const manifest = result.manifest;

    // Assert
    expect(manifest).toMatchSnapshot();
  });
});
