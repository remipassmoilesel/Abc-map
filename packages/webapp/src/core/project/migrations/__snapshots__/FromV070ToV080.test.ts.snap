// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromV070ToV080 migrate should work 1`] = `
Object {
  "layers": Array [
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "© OpenStreetMap contributors.",
        ],
        "id": "3baf449f-83d7-4ee4-bf9b-66ab377c9d3a",
        "model": "OSM",
        "name": "OpenStreetMap",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "features": Object {
        "features": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                483537.70856973104,
                5384734.194694325,
              ],
              "type": "Point",
            },
            "id": "PKNy6gKXSc",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(147,18,49,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                462762.70856973104,
                5390434.194694325,
              ],
              "type": "Point",
            },
            "id": "G0_XyWz64a",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(147,18,49,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                474012.70856973104,
                5385034.194694325,
              ],
              "type": "Point",
            },
            "id": "yOHGhoTseI",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(18,19,147,0.9)",
              "abc:style:point:icon": "twbs/0_square-fill.inline.svg",
              "abc:style:point:size": 40,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                Array [
                  Array [
                    494187.70856973104,
                    5392909.194694325,
                  ],
                  Array [
                    485262.70856973104,
                    5390959.194694325,
                  ],
                  Array [
                    482637.70856973104,
                    5401984.194694325,
                  ],
                  Array [
                    496362.70856973104,
                    5405884.194694325,
                  ],
                  Array [
                    499512.70856973104,
                    5391934.194694325,
                  ],
                  Array [
                    494187.70856973104,
                    5392909.194694325,
                  ],
                ],
              ],
              "type": "Polygon",
            },
            "id": "JNI5QMnOHS",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:fill:color1": "#cc8d70",
              "abc:style:fill:color2": "rgba(255,255,255,0.60)",
              "abc:style:fill:pattern": "abc:style:fill:hatching:oblique:right",
              "abc:style:stroke:color": "rgba(18,90,147,0.60)",
              "abc:style:stroke:width": 3,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                472362.70856973104,
                5396659.194694325,
              ],
              "type": "Point",
            },
            "id": "NXvMFHnhSk",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(147,18,49,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                Array [
                  518412.70856973104,
                  5390209.194694325,
                ],
                Array [
                  507687.70856973104,
                  5377984.194694325,
                ],
                Array [
                  513387.70856973104,
                  5370109.194694325,
                ],
                Array [
                  526812.708569731,
                  5377684.194694325,
                ],
                Array [
                  504762.70856973104,
                  5393809.194694325,
                ],
              ],
              "type": "LineString",
            },
            "id": "f9ToQfhmws",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:stroke:color": "rgba(18,90,147,0.60)",
              "abc:style:stroke:width": 3,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                478137.70856973104,
                5378659.194694325,
              ],
              "type": "Point",
            },
            "id": "gmbpVxddjV",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(147,18,49,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                488862.70856973104,
                5383534.194694325,
              ],
              "type": "Point",
            },
            "id": "Ydq46oceSH",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(18,19,147,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
              "abc:style:text:value": "",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                498912.70856973104,
                5377459.194694325,
              ],
              "type": "Point",
            },
            "id": "LneUQyncso",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(147,18,49,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
              "abc:style:text:value": "",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                501837.70856973104,
                5381734.194694325,
              ],
              "type": "Point",
            },
            "id": "C2NdE0ZmyU",
            "properties": Object {
              "abc:feature:selected": false,
              "abc:style:point:color": "rgba(18,19,147,0.9)",
              "abc:style:point:icon": "twbs/0_star-fill.inline.svg",
              "abc:style:point:size": 20,
            },
            "type": "Feature",
          },
        ],
        "type": "FeatureCollection",
      },
      "metadata": Object {
        "active": true,
        "id": "81775cd7-1e77-4c07-ac37-e6989a9b229f",
        "name": "Géométries",
        "opacity": 1,
        "type": "Vector",
        "visible": true,
      },
      "type": "Vector",
    },
  ],
  "layouts": Array [
    Object {
      "format": Object {
        "height": 210,
        "id": "A4_LANDSCAPE",
        "orientation": "landscape",
        "width": 297,
      },
      "id": "f3418195-aa3c-4add-870c-a3e09dc33e0a",
      "name": "Page 1",
      "textFrames": Array [
        Object {
          "content": Array [
            Object {
              "children": Array [
                Object {
                  "text": "",
                },
              ],
              "type": "paragraph",
            },
            Object {
              "border": 0,
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "LineString",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {},
                                "stroke": Object {},
                                "text": Object {},
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 1",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(18,19,147,0.9)",
                                  "icon": "twbs/0_star-fill.inline.svg",
                                  "size": 12,
                                },
                                "stroke": Object {},
                                "text": Object {
                                  "value": "",
                                },
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 2",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(147,18,49,0.9)",
                                  "icon": "twbs/0_star-fill.inline.svg",
                                  "size": 20,
                                },
                                "stroke": Object {},
                                "text": Object {
                                  "value": "",
                                },
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 3",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(18,19,147,0.9)",
                                  "icon": "twbs/0_star-fill.inline.svg",
                                  "size": 20,
                                },
                                "stroke": Object {},
                                "text": Object {
                                  "value": "",
                                },
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 4",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(18,19,147,0.9)",
                                  "icon": "twbs/0_square-fill.inline.svg",
                                  "size": 40,
                                },
                                "stroke": Object {},
                                "text": Object {},
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 5",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
              ],
              "type": "table",
            },
            Object {
              "children": Array [
                Object {
                  "text": "",
                },
              ],
              "type": "paragraph",
            },
          ],
          "id": "###-FAKE-NANOID-1-###",
          "position": Object {
            "x": 0,
            "y": 0,
          },
          "size": Object {
            "height": 517.5,
            "width": 392,
          },
        },
      ],
      "view": Object {
        "center": Array [
          479684.677727075,
          5377825.024768071,
        ],
        "projection": Object {
          "name": "EPSG:3857",
        },
        "resolution": 46.966404979,
      },
    },
  ],
  "metadata": Object {
    "containsCredentials": false,
    "id": "224855c1-7085-4410-abcc-5b3d845c7ad0",
    "name": "Projet du 15/03/2022",
    "public": true,
    "version": "0.8.0",
  },
  "sharedViews": Array [
    Object {
      "id": "VxtjLCKzWPXB1g2CSjLir",
      "layers": Array [
        Object {
          "layerId": "3baf449f-83d7-4ee4-bf9b-66ab377c9d3a",
          "visible": true,
        },
        Object {
          "layerId": "81775cd7-1e77-4c07-ac37-e6989a9b229f",
          "visible": true,
        },
      ],
      "textFrames": Array [
        Object {
          "content": Array [
            Object {
              "children": Array [
                Object {
                  "text": "",
                },
              ],
              "type": "paragraph",
            },
            Object {
              "border": 0,
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "LineString",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {},
                                "stroke": Object {},
                                "text": Object {},
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 1",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(18,19,147,0.9)",
                                  "icon": "twbs/0_star-fill.inline.svg",
                                  "size": 12,
                                },
                                "stroke": Object {},
                                "text": Object {
                                  "value": "",
                                },
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 2",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(147,18,49,0.9)",
                                  "icon": "twbs/0_star-fill.inline.svg",
                                  "size": 20,
                                },
                                "stroke": Object {},
                                "text": Object {
                                  "value": "",
                                },
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 3",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(18,19,147,0.9)",
                                  "icon": "twbs/0_star-fill.inline.svg",
                                  "size": 20,
                                },
                                "stroke": Object {},
                                "text": Object {
                                  "value": "",
                                },
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 4",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "align": "center",
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "text": "",
                                },
                              ],
                              "geometryType": "Point",
                              "style": Object {
                                "fill": Object {},
                                "point": Object {
                                  "color": "rgba(18,19,147,0.9)",
                                  "icon": "twbs/0_square-fill.inline.svg",
                                  "size": 40,
                                },
                                "stroke": Object {},
                                "text": Object {},
                              },
                              "type": "map-symbol",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "text": "Entry 5",
                            },
                          ],
                          "type": "paragraph",
                        },
                      ],
                      "type": "table-cell",
                    },
                  ],
                  "type": "table-row",
                },
              ],
              "type": "table",
            },
            Object {
              "children": Array [
                Object {
                  "text": "",
                },
              ],
              "type": "paragraph",
            },
          ],
          "id": "###-FAKE-NANOID-2-###",
          "position": Object {
            "x": 0,
            "y": 0,
          },
          "size": Object {
            "height": 517.5,
            "width": 392,
          },
        },
      ],
      "title": "Vue principale",
      "view": Object {
        "center": Array [
          500937.708569731,
          5388389.272819325,
        ],
        "projection": Object {
          "name": "EPSG:3857",
        },
        "resolution": 75,
      },
    },
  ],
  "view": Object {
    "center": Array [
      500937.708569731,
      5388389.272819325,
    ],
    "projection": Object {
      "name": "EPSG:3857",
    },
    "resolution": 75,
  },
}
`;
